xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-filters/xwiki-platform-notifications-filters-default/src/main/java/org/xwiki/notifications/filters/migration/WatchListObjectsRemovalTaskConsumer.java

```diff
@@ -0,0 +1,100 @@
+import org.xwiki.index.TaskConsumer;
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.model.reference.EntityReference;
+import org.xwiki.model.reference.LocalDocumentReference;
+
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.XWikiException;
+import com.xpn.xwiki.doc.XWikiDocument;
+import com.xpn.xwiki.objects.BaseObject;
+
+/**
+ * Task for removing WatchListClass xobjects from page. This task should only be triggered when
+ * {@link R160000000XWIKI17243DataMigration} has been done.
+ * Note that this task might also create an autowatch xobject to get the new property if it's not defined, as we use
+ * to fallback on the autowatch property defined in the WatchListClass xobject.
+ *
+ * @version $Id$
+ * @since 16.0.0RC1
+ */
+@Component
+@Named(WatchListObjectsRemovalTaskConsumer.TASK_NAME)
+@Singleton
+public class WatchListObjectsRemovalTaskConsumer implements TaskConsumer
+{
+    static final String TASK_NAME = "watchlist-xobject-removal";
+
+    private static final String XWIKI_SPACE = "XWiki";
+
+    private static final List<String> NOTIFICATIONS_CODE_SPACE = Arrays.asList(XWIKI_SPACE, "Notifications", "Code");
+
+    private static final LocalDocumentReference CLASS_REFERENCE =
+        new LocalDocumentReference(XWIKI_SPACE, "WatchListClass");
+
+    private static final LocalDocumentReference AUTOMATIC_WATCH_CLASS_REFERENCE =
+        new LocalDocumentReference(NOTIFICATIONS_CODE_SPACE, "AutomaticWatchModeClass");
+
+    @Inject
+    private Provider<XWikiContext> contextProvider;
+
+    @Override
+    public void consume(DocumentReference documentReference, String version) throws IndexException
+    {
+        XWikiContext context = this.contextProvider.get();
+        EntityReference classReference = CLASS_REFERENCE.appendParent(documentReference.getWikiReference());
+        EntityReference autoWatchClassReference =
+            AUTOMATIC_WATCH_CLASS_REFERENCE.appendParent(documentReference.getWikiReference());
+        try {
+            XWikiDocument document = context.getWiki().getDocument(documentReference, context);
+
+            // We use to fallback on the old WatchClass xobject automaticwatch property when it was defined, so we
+            // also take back this value and create the new autowatch xobject if needed.
+            BaseObject watchXObject = document.getXObject(classReference);
+            if (watchXObject != null && document.getXObject(autoWatchClassReference) == null) {
+                String automaticwatch = watchXObject.getStringValue("automaticwatch");
+                if (!StringUtils.isBlank(automaticwatch) && !"default".equals(automaticwatch)) {
+                    BaseObject autowatchXObject = document.newXObject(autoWatchClassReference, context);
+                    autowatchXObject.setStringValue("automaticWatchMode", automaticwatch);
+                }
+            }
+
+            if (document.removeXObjects(classReference)) {
```

Comment:
Is it expected that the new XObject created line 88 won't be saved if `removeXObjects` returns false?

Reply:
`removeXObjects` returns false only if there's no object with the given reference, so the if `watchXObject != null` would have been true. So basically you cannot enter in previous if without entering in that if too

Reply:
The if might be superfluous in this case?

Reply:
Which one? I need the first if, since it's really an edge case. This if is less likely to return false, since in theory I enter the TaskConsumer if I have an XWikiListClass xobject to remove, now since the method does return a boolean better safe and use the result for saving IMO.